src/core/main/Reflection/InstanceAnalyzer.cs

```diff
@@ -0,0 +1,274 @@
+                                }
+                            }
+                        
+                            context.BreadcrumbStack.Pop();
+                        }
+                    }
+                    else
+                    {
+                        var enumerable = value as IEnumerable;
+
+                        if (enumerable == null)
+                        {
+                            throw new InvalidOperationException($"Trying to iterate over {context.BreadcrumbAsString}.{memberInfo.Name}, but could not cast to IEnumerable.");
+                        }
+
+                        var index = 0;
+                        foreach (var element in enumerable)
+                        {
+                            context.BreadcrumbStack.Push($"{memberInfo.Name}[{index}]");
+                            CallListener(listener, context, memberInfo, () => element);
+
+                            if (ShouldRecurse(element.GetType()))
+                            {
+                                if (context.CanGoDeeper())
+                                {
+                                    Worker(element, listener, context);
+                                }
+                                else
+                                {
+                                    listener.OnMaxDepth(context);
+                                }
+                            }
+                        
+                            context.BreadcrumbStack.Pop();
+                            index++;
+                        }
+                    }
+                }
+            }
+
+            else if (ShouldRecurse(valueType))
+            {
+                //
+                // recursion
+                //
+                if (context.CanGoDeeper())
+                {
+                    object value = valueGetter.Invoke();
+                    
+                    if (value != null)
+                    {
+                        context.BreadcrumbStack.Push(memberInfo.Name);
+                        Worker(value, listener, context);
+                        context.BreadcrumbStack.Pop();
+                    }
+                }
+                else
+                {
+                    listener.OnMaxDepth(context);
+                }
+            }
+        }
+
+        /// <summary>
+        /// A quick lookup of types that we do not want to recurse.
+        ///
+        /// The value part is not used and means nothing
+        /// </summary>
+        private readonly Dictionary<Type, bool> nonRecursedTypes = new Dictionary<Type, bool>
+        {
+            { typeof(char), true },
+            { typeof(bool), true },
+            { typeof(byte), true },
+            { typeof(short), true },
+            { typeof(int), true },
+            { typeof(long), true },
+            { typeof(float), true },
+            { typeof(double), true },
+            { typeof(decimal), true },
```

Comment:
How about nullable value types, they are also non recursable I assume?

Reply:
We should check for nullable types yes, and also streams.